psdisk/psdisk.c

```diff
@@ -12,156 +12,570 @@
-	printf("\n%s %-35s %-30s", "   - h ", " ", "= Show help");
+	printf("\n\t%s -h", app);
+	printf("\n\t%s <image-path> -m <mem-size,sector-size>", app);
+	printf("\n\t%s <image-path> -m <mem-size,sector-size> [options]", app);
+	printf("\n\nOptions:");
+	printf(HELP_ALIGMENT, "   - m ", "<mem-size,sector-size>", "declare memory's parameters");
+	printf(HELP_ALIGMENT, "   - p ", "<name,offset,size,type>", "declare partition");
+	printf(HELP_ALIGMENT, "   - r ", "<name>", "remove partition");
+	printf(HELP_ALIGMENT, "   - o ", " ", "partition table is located in the last sector of memory");
+	printf(HELP_ALIGMENT, "   - h ", " ", "show help");
+	printf("\n");
+	printf("\nPartition types: \n");
+	printf("\t- meterfs = 0x75,\n");
+	printf("\t- raw = 0x51.\n");
 	printf("\n\n");
 }
 
 
-static ssize_t psdisk_writeData(unsigned int addr, const void *buff, size_t bufflen)
+static void psdisk_showPartsTable(void)
+{
+	int sz;
+	struct plist_node_t *node;
+
+	fseek(psdisk_common.file, 0L, SEEK_END);
+	sz = ftell(psdisk_common.file);
+
+	printf("\n");
+	printf(BOLDWHITE "File system %s: %d bytes\n"RESET, psdisk_common.fileName, sz);
+	printf("Memory size %u bytes\n", psdisk_common.mem.memSize);
+	printf("Units: sector %u bytes \n", psdisk_common.mem.sectorSize);
+	printf("\n");
+	printf(BOLDWHITE PARTITION_HEADER_ALIGMENT RESET, "Name", "Start", "End", "Sectors", "Size", "Type");
+	LIST_FOREACH(node, &psdisk_common.list, ptrs) {
+		printf(PARTITION_DATA_ALIGMENT, node->pHeader.name,
+										node->pHeader.offset,
+										node->pHeader.offset + node->pHeader.size,
+										node->pHeader.size / psdisk_common.mem.sectorSize,
+										node->pHeader.size,
+										psdisk_getTypeName(node->pHeader.type));
+	}
+	printf("\n");
+}
+
+
+/* Callbacks to ptable library */
+
+static ssize_t psdisk_readImg(unsigned int addr, void *buff, size_t bufflen)
+{
+	int offs;
+
+	if (PTABLE_WITH_OFFSET(psdisk_common.opts))
+		offs = addr;
+	else
+		offs = addr - (psdisk_common.mem.memSize - psdisk_common.mem.sectorSize);
+
+	if (fseek(psdisk_common.file, offs, SEEK_SET) != 0) {
+		printf("Cannot change file position, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	return fread(buff, sizeof(uint8_t), bufflen, psdisk_common.file);
+}
+
+
+static ssize_t psdisk_writeImg(unsigned int addr, const void *buff, size_t bufflen)
+{
+	ssize_t res;
+	char *wBuff;
+
+	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
+		printf("Cannot change file position, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	if ((wBuff = (char *)malloc(psdisk_common.mem.sectorSize)) == NULL) {
+		printf("Cannot allocate memory, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	memset(wBuff, 0xff, psdisk_common.mem.sectorSize);
+
+	if (PTABLE_WITH_OFFSET(psdisk_common.opts)) {
+		while (addr > 0) {
+			fwrite(wBuff, sizeof(uint8_t), psdisk_common.mem.sectorSize, psdisk_common.file);
+			addr -= psdisk_common.mem.sectorSize;
+		}
+	}
+
+	memcpy(wBuff, buff, bufflen);
```

Comment:
Is there a possibility that bufflen > sectorSize? If so it might turn out ugly

Reply:
Do we have guarantee that bufflen is less or equal sectorSize?

Reply:
Condition to check this case was added. If bufflen exceeds sector size, partition table is not written to file.